db/multi_cf_iterator_impl.cc

```diff
@@ -0,0 +1,51 @@
+//  Copyright (c) Meta Platforms, Inc. and affiliates.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "db/multi_cf_iterator_impl.h"
+
+#include <cassert>
+
+namespace ROCKSDB_NAMESPACE {
+
+void MultiCfIteratorImpl::SeekToFirst() {
+  Reset();
+  assert(cfhs_.size() == iterators_.size());
+  int i = 0;
+  for (auto& iter : iterators_) {
+    auto& cfh = cfhs_[i];
+    iter->SeekToFirst();
+    if (iter->Valid()) {
+      assert(iter->status().ok());
+      min_heap_.push(MultiCfIteratorInfo{iter, cfh, i});
+    } else {
+      considerStatus(iter->status());
+    }
+    ++i;
+  }
+}
+
+void MultiCfIteratorImpl::Next() {
+  assert(Valid());
+
+  auto* current = min_heap_.top().iterator;
+  std::string current_key_copy =
```

Comment:
Here, we first store the key from the iterator at the top of the heap. Then the top iterator moves to the next element and heap gets re-heapified. If the same key appears at the top of the heap, it repeats. Basically, this Next() finds the next key in order across all the iterators in the heap by skipping the same key that we've seen.

I'd love to see if there's a better way to achieve this (e.g. without copying the first key)

Reply:
It seems like you should be able to keep one iterator at the current key (remove from heap), use that key Slice to ensure all others have iterated past it (until top of heap is past it), then advance the first iterator past that key and re-place on the heap.